A build and deploy script to manage cluster federation lifecycle and a Makefile wrapper to drive that process.

madhusudancs opened the pull request:
Also includes a sample config file to describe clusters.

The build script implements the following things:
1. Generates the required configs.
2. Builds the hyperkube binary and the corresponding docker image.
3. Pushes the image to a specified repository.
4. Pulls the federation installer docker images.
5. Builds the Kubernetes clusters described the config.json file.
6. Pushes the federation components to one of the Kubernetes clusters
   built in the previous step.
7. Also turns down the federation components and the Kubernetes
   clusters.

**NOTE**: Installer images are right now being pulled from my public repository of docker images. I am working on pushing them to our release repository.

``` release-note
Cluster Federation components can now be built and deployed using the make command. Please see federation/README.md for details.
```

cc @kubernetes/sig-cluster-federation @colhom 

Fixes: Issue #26655

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29515)

<!-- Reviewable:end -->

madhusudancs committed:
Implement a build and deploy script to turn up/down federation.

Also, wrap the script around a Makefile. And also provide a sample
config file to describe clusters.

The build script implements the following things:
1. Generates the required configs.
2. Builds the hyperkube binary and the corresponding docker image.
3. Pushes the image to a specified repository.
4. Pulls the federation installer docker images.
5. Builds the Kubernetes clusters described the config.json file.
6. Pushes the federation components to one of the Kubernetes clusters
   built in the previous step.
7. Also turns down the federation components and the Kubernetes
   clusters.

madhusudancs committed:
Add some status logging.

madhusudancs committed:
Installer images are supposed to be public, so we can directly pull without gcloud prefix.

madhusudancs committed:
Build both kubectl and hypekube together, one is required for the other.

madhusudancs committed:
Recompute KUBE_VERSION after rebuild in the dev environment.

madhusudancs committed:
Add a function to ensure that required credential files exist.

madhusudancs committed:
Add a README.md file for cluster federation.

madhusudancs committed:
Add detect-project capability.

colhom commented:
@madhusudancs this is awesome!!! I feel like a human can **actually understand** what's going on now! 

I will give this a thorough review later today. Just wanted to say how happy this makes me.

madhusudancs commented:
@colhom thanks!

colhom commented:
@madhusudancs if I'm reading this correctly- we don't really need the `federation/cluster/*.sh` scripts anymore? Maybe we should just replace them in this PR

madhusudancs commented:
@colhom you are right. However, kubernetes-anywhere, on which this depends, does not work very well with AWS yet. I think it is a good idea to leave the `cluster/*.sh` scripts as is until we are fairly confident that it works with AWS as well, i.e. we have tested out this thing well on AWS.

colhom commented:
@madhusudancs what aws-specific issues are you running into? i see you're using the kube-up bash libraries, so there shouldn't (theoretically) be any provider-specific problems.

madhusudancs commented:
@colhom Although it is theoretically possible to use kube-up with this script, this PR uses [kubernetes-anywhere](https://github.com/kubernetes/kubernetes-anywhere) to turn up clusters. 

I don't remember the exact error and I tried it only once. When I tried to bring up the cluster with [kubernetes-anywhere](https://github.com/kubernetes/kubernetes-anywhere), the AWS resources were all created correctly but I could still not reach kube-apiserver. I did not spend time debugging. But any help there is welcome.

madhusudancs committed:
Docs munger and flags verification fixes.

mml commented on federation/Makefile:
```diff
@@ -0,0 +1,26 @@
+
+.PHONY: build
+build:
+	./build.sh $(do)
```
Any reason not to make these individual make targets?  `make gen`, `make deploy`, `make destroy`

madhusudancs replied:
This is one of the things that I am going to work on next, unless someone offers to help (as listed in the TODO in the README file). I started out this way because it was easier to experiment with things if I had to make changes to only one script (build.sh).

mml commented on federation/Makefile:
```diff
@@ -0,0 +1,26 @@
+# limitations under the License.
+
+DBG_MAKEFILE ?=
+ifeq ($(DBG_MAKEFILE),1)
```
Does this work with BSD make?  The most likely place that would matter is on a Mac, though I'm not sure which make it comes with.  I can check myself.

madhusudancs replied:
Not entirely sure. Assuming FreeBSD actually ships BSD make, this should work - http://www.freebsd.org/cgi/man.cgi?make(1). I don't have access to a Mac, so I need your help here.

Also, this isn't specific to federation, the top-level Makefile (the one in the root directory of the source) also has this, so I think this should work on Mac.

mml commented on federation/README.md:
```diff
@@ -0,0 +1,61 @@
+
+Please see the [Kubernetes Development Guide](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md)
+for initial setup. Once you have the development environment setup
+as explained in that guide, you also need to install [`jq`](https://stedolan.github.io/jq/download/)
```
I like `jq` and use it, too, but it seems like an inconvenient hurdle to require it just to try out federation.  Are we doing anything that we couldn't accomplish with `sed`?

madhusudancs replied:
Thanks for starting this conversation. I meant to debate about this but somehow missed it while sending this PR.

My argument to myself so far has been that:

1. `jq` is really easy to install. It is in the upstream repositories of popular Linux distributions. No additional repositories needed, just `apt-get install jq`, `dnf install jq` and so on. There are packages and/or binaries readily available for Mac, Windows and other platforms - https://stedolan.github.io/jq/download/. So it is not that much of a hassle IMO.
2. This is only dev environments. We will not require `jq` in the shipped release/distribution.
3. It is not doing something special that we cannot accomplish with `sed`, but the intent is so much clear with `jq`. Also, it guarantees that it modifies only that particular JSON object we intend to modify and not accidentally modify some other field which matches the regexp.

Please let me know your thoughts. Open for suggestions.

nikhiljindal commented on federation/README.md:
````diff
@@ -0,0 +1,61 @@
+Building cluster federation should be as simple as running:
+
+```shell
+make build do=gen
````
Why do=gen? Is do=build better?

madhusudancs replied:
There will be neither a do=gen or do=build very soon. See the TODO list below. And also my reply to @mml's comment https://github.com/kubernetes/kubernetes/pull/29515#discussion_r73233835.

For a bit of history, `gen`, 'deploy` and `destroy` were the phrases that kubernetes-anywhere used for the different phases of the lifecycle. Since this work is based on the work in kubernetes-anywhere, it started with the same phrases for ease/simplicity.

nikhiljindal commented on federation/build.sh:
```diff
@@ -0,0 +1,211 @@
+
+function ensure_files() {
+	kube::log::status "Ensure credential files exist..."
+	if [[ ! -f "${GOOGLE_APPLICATION_CREDENTIALS}" ]]; then
```
Should check this only if CLOUD_PROVIDER=gce or gke?
If this script is not expected to work for other cloud providers, we should probably error out earlier with a clear error message.

madhusudancs replied:
Done. 

Also, a side note. `gce` is the only supported provider in this very first version. I want to get the first version in before I can improve it. I don't want to be continuously perfecting this in a vacuum. Improvements are going to land in follow up PRs.

nikhiljindal commented on federation/README.md:
````diff
@@ -0,0 +1,61 @@
+```
+
+To deploy clusters and install federation components, edit the
+`config.default.json` file to describe your clusters and run
````
Which cluster will the federation components be installed on? First/last one in the json?

madhusudancs replied:
It can be any of the clusters. It just installs the federation components in the cluster which is the default context in the kubeconfig. But to make it more predictable I set the first cluster as the default context - https://github.com/madhusudancs/kubernetes-anywhere/blob/federation/federation/entrypoint.sh#L52

nikhiljindal replied:
Will be good to mention this in the README

nikhiljindal commented on federation/config.default.json:
```diff
@@ -0,0 +1,83 @@
+[
+  {
```
curious where is the schema for this defined?

madhusudancs replied:
The schema is here - https://github.com/kubernetes/kubernetes-anywhere/blob/master/Kconfig. You need to follow the source-chain.

There is a discussion to move to a more explicit and simpler schema and/or config generation mechanism. See - https://github.com/kubernetes/kubernetes-anywhere/issues/128

nikhiljindal replied:
How about putting that link in the README file?

madhusudancs committed:
Ensure the provider is supported and exit otherwise.

madhusudancs commented:
@mml @nikhiljindal PTAL.

mml commented:
Reviewed 2 of 4 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[federation/build.sh, line 74 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-KOqU1FXl8rYyjQ3NNth:-KOqU1FXl8rYyjQ3NNti:b7qqaax) ([raw file](https://github.com/kubernetes/kubernetes/blob/368fad710a4ec4f866befb0423e8fe84a32ac7d0/federation/build.sh#L74)):*
> ```Shell
> 
> function cleanup {
> 	rm -rf "${TMP_DIR}"
> ```

Tab literals here (through the whole file).  Not sure we have a style guide, but it seems to me that 2 blank spaces is what I've seen, except in go files.

---

*[federation/Makefile, line 16 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-KOgn3EEqHxCNYV7sMRH-r2-16:-KOqUQhsBh0eABcmTjB2:bqxw4yq) ([raw file](https://github.com/kubernetes/kubernetes/blob/0454a81017e890a474b96a8d1408e7bcde3d24f2/federation/Makefile#L16)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Not entirely sure. Assuming FreeBSD actually ships BSD make, this should work - http://www.freebsd.org/cgi/man.cgi?make(1). I don't have access to a Mac, so I need your help here.
> 
> 
> 
> Also, this isn't specific to federation, the top-level Makefile (the one in the root directory of the source) also has this, so I think this should work on Mac.
> 
> 
</details>
OK.  I checked and it's actually GNU make on Mac.

---

*[federation/README.md, line 12 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-KOgn3EEqHxCNYV7sMRI-r2-12:-KOqWzj33jZqIrJxc6ry:bj1yeeo) ([raw file](https://github.com/kubernetes/kubernetes/blob/0454a81017e890a474b96a8d1408e7bcde3d24f2/federation/README.md#L12)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Thanks for starting this conversation. I meant to debate about this but somehow missed it while sending this PR.
> 
> 
> 
> My arguments to self so far:
> 
> 
> 
> 1. `jq` is really easy to install. It is in the upstream repos of popular Linux distributions. No additional repositories needed, just `apt-get install jq`, `dnf install jq` and so on. There are packages and/or binaries readily available for Mac, Windows and other platforms - https://stedolan.github.io/jq/download/. So it is not that much of a hassle IMO.
> 
> 2. And it is required only for dev environments. We will not require `jq` in the shipped release/distribution.
> 
> 3. It is not doing something special that we cannot accomplish with `sed`, but the intent is so much clear with `jq`. Also, it guarantees that it modifies only that particular JSON object we intend to modify and not accidentally modify some other field which matches the regexp.
> 
> 
> 
> Please let me know your thoughts. Open to suggestions.
</details>
The only thing I have to add is that we should do this however kubernetes anywhere does it.  From a glance, it looks like they use jsonnet as a way to templatize json.

If you want to check this in with jq, that is fine.  Just please leave a TODO comment to revisit this and rationalize it with KA.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515)*
<!-- Sent from Reviewable.io -->

madhusudancs committed:
Add a TODO to re-evaluate jq dependency.

madhusudancs committed:
Switch tab indentation to two spaces.

madhusudancs commented:
@mml Addressed review comments. PTAL.

---

Review status: 2 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/build.sh, line 74 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-KOqU1FXl8rYyjQ3NNth:-KOsN2blO5eZwRFoOBQZ:bam0u8z) ([raw file](https://github.com/kubernetes/kubernetes/blob/368fad710a4ec4f866befb0423e8fe84a32ac7d0/federation/build.sh#L74)):*
<details><summary><i>Previously, mml (Matt Liggett) wrote…</i></summary>
> Tab literals here (through the whole file).  Not sure we have a style guide, but it seems to me that 2 blank spaces is what I've seen, except in go files.
</details>
Switched to two-spaces.

---

*[federation/README.md, line 12 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-KOgn3EEqHxCNYV7sMRI-r2-12:-KOqZqjvFCYLgnWIUTam:bu9sq6x) ([raw file](https://github.com/kubernetes/kubernetes/blob/0454a81017e890a474b96a8d1408e7bcde3d24f2/federation/README.md#L12)):*
<details><summary><i>Previously, mml (Matt Liggett) wrote…</i></summary>
> The only thing I have to add is that we should do this however kubernetes anywhere does it.  From a glance, it looks like they use jsonnet as a way to templatize json.
> 
> If you want to check this in with jq, that is fine.  Just please leave a TODO comment to revisit this and rationalize it with KA.
</details>
KA also has `jq` dependency today. I have left a TODO anyway, saying we need to re-evaluate using `jq`.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-:-KOsNQZSyOKmSju3YCKX:blte4z1)*
<!-- Sent from Reviewable.io -->

mml commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[federation/README.md, line 12 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-KOgn3EEqHxCNYV7sMRI-r2-12:-KP-EoSQWorpmJihpZrT:bnl9g4s) ([raw file](https://github.com/kubernetes/kubernetes/blob/0454a81017e890a474b96a8d1408e7bcde3d24f2/federation/README.md#L12)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> KA also has `jq` dependency today. I have left a TODO anyway, saying we need to re-evaluate using `jq`.
</details>
OK SGTM

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29515#-:-KP-Ey14cuEpNKz3mvQ8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@k8s-bot test this issue: #IGNORE

nikhiljindal commented on federation/README.md:
````diff
@@ -0,0 +1,67 @@
+`config.default.json` file to describe your clusters and run
+
+```shell
+make build do=deploy
````
Is do=gen a required step or can I directly run deploy if I want to deploy release images.
Also how can I specify the image that I want to use for federation (if I want to use a release image)?

(Am just asking for my curiosity. Its fine to say we dont support it right now, if thats the case).

nikhiljindal commented:
My comments are all optional and mostly questions for my own curiosity. Dont need to block this PR on those.
